embassy-stm32/src/gpio.rs

```diff
@@ -521,32 +530,46 @@ impl<'d> OutputOpenDrain<'d> {
 }
 
 /// GPIO output type
+#[derive(Debug, Copy, Clone)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
 pub enum OutputType {
     /// Drive the pin both high or low.
     PushPull,
     /// Drive the pin low, or don't drive it at all if the output level is high.
     OpenDrain,
 }
 
-impl From<OutputType> for AFType {
-    fn from(value: OutputType) -> Self {
-        match value {
-            OutputType::OpenDrain => AFType::OutputOpenDrain,
-            OutputType::PushPull => AFType::OutputPushPull,
+#[cfg(gpio_v1)]
+impl From<OutputType> for vals::CnfOut {
+    fn from(output_type: OutputType) -> Self {
+        match output_type {
+            OutputType::PushPull => vals::CnfOut::ALTPUSHPULL,
+            OutputType::OpenDrain => vals::CnfOut::ALTOPENDRAIN,
+        }
+    }
+}
+
+#[cfg(gpio_v2)]
+impl From<OutputType> for vals::Ot {
+    fn from(output_type: OutputType) -> Self {
+        match output_type {
+            OutputType::PushPull => vals::Ot::PUSHPULL,
+            OutputType::OpenDrain => vals::Ot::OPENDRAIN,
         }
     }
 }
 
-/// Alternate function type settings
+/// Alternate function type settings.
 #[derive(Debug, Copy, Clone)]
 #[cfg_attr(feature = "defmt", derive(defmt::Format))]
-pub enum AFType {
-    /// Input
-    Input,
-    /// Output, drive the pin both high or low.
-    OutputPushPull,
-    /// Output, drive the pin low, or don't drive it at all if the output level is high.
-    OutputOpenDrain,
+pub enum AfType {
+    /// Input with optional pullup or pulldown.
+    Input(Pull),
+    /// Output with output type and speed and no pull-up or pull-down.
+    Output(OutputType, Speed),
```

Comment:
`AfType::Output(_, _)` is the same as `AfType::OutputPull(_, _, Pull::None)`

it's a bit strange to have "redundancy" in enums like this, and i think this will generate extra code in functions that take an `AfType` even if it's not used. It's probably better to:

- make a single `Output` case in all versions.
- The `Output` case contains `OutputType, Speed` on v1, and `OutputType, Speed, Pull` on v2.
- since it's not nice that the struct changes depending on gpio version (making it hard for the user to write code that works on both versions), perhaps we can make the enum private, expose an opaque `pub struct AfType(AfTypeInner)`, and add methods to construct them. so `fn output(type, speed)` is available on all versions and `fn output_pull(type, speed, pull)` only on v2, but they both return the same enum case.

Reply:
The reason why I chose to introduce a conditional `OutputPull` variant instead of a conditional field in the `Output` variant was to minimize the number of `#[cfg]`-s for users while keeping the code as simple as possible (no need for a wrapped private type and/or constructor methods). Note that `AfType` is by design matched exactly in once place, in the `SealedPin::set_as_af()` method, so the cost of extra enum variant in terms of _lines_ of code is quite small.

However, if you are worried about the cost in _bytes_ of code in flash, that's another matter. In the happy case, where `set_as_af()` is called with a statically known pin and a statically known `AfType`, the compiler should be able to reduce everything to a constant a produce the optimal code. But if the compiler is not able to optimize out the `AfType` enum, the code size might suffer, because `set_as_af()` will need a lot of code to convert all the enums to register values. This is potentially made worse by `set_as_af()` being a trait method and marked as `#[inline]`, so we could end up with many copies of the same code in flash.

I will need to inspect the size of generated code to verify my assumptions, but here are some options how we could minimize the size of the generated code:

1. Move the code of `set_as_af()` into a function outside of the `SealedPin` trait and don't mark it as `#[inline]`: this means that we will have just a single copy of this code in flash, so its size would not be that important, and every usage of `set_as_af()` should compile to a function call, which should be quite small (probably smaller than the most optimized inline code, but I need to verify that)
2. Change `AfType` from an enum to a struct with private fields, and store the register values directly in this struct (e.g. for gpio_v2, the struct would contain fields `pupdr`, `ot` and `ospeedr`), and introduce constructor functions like `AfType::input()` and `AfType::output_pull()` to create this struct from user-friendly enums like `Pull` and `Speed`. This would move all the `match`-es out of the `set_as_af()` function into the constructor functions, where the compiler should be able to almost always optimize them out.

I will do some experiments and get back with the results.